sam_handler/handler.py

```diff
@@ -0,0 +1,247 @@
+        SAM CLI.
+
+        The environment variables dict created by this method will inherit all existing
+        environment variables in the current environment, but the AWS connection environment
+        variables will be saved overridden by the ones for this stack.
+
+        Returns:
+            The dictionary of environment variables.
+        """
+        envs = self.environment_variables.copy()
+        # Set aws environment variables specific to whatever AWS configuration has been set on the
+        # stack's connection manager.
+        credentials: Credentials = self.connection_manager._get_session(
+            self.connection_manager.profile,
+            self.connection_manager.region,
+            self.connection_manager.iam_role
+        ).get_credentials()
+        envs.update(
+            AWS_ACCESS_KEY_ID=credentials.access_key,
+            AWS_SECRET_ACCESS_KEY=credentials.secret_key,
+        )
+
+        # There might not be a session token, so if there isn't one, make sure it doesn't exist in
+        # the envs being passed to the subprocess
+        if credentials.token is None:
+            envs.pop('AWS_SESSION_TOKEN', None)
+        else:
+            envs['AWS_SESSION_TOKEN'] = credentials.token
+
+        return envs
+
+
+class SAM(TemplateHandler):
+    """A template handler for AWS SAM templates. Using this will allow Sceptre to work with SAM to
+    build and package a SAM template and deploy it with Sceptre.
+    """
+
+    SAM_ARTIFACT_DIRECTORY = 'sam_artifacts'
+
+    def __init__(
+        self,
+        name,
+        arguments=None,
+        sceptre_user_data=None,
+        connection_manager=None,
+        stack_group_config=None,
+        *,
+        invoker_class=SamInvoker,
+        get_temp_dir=tempfile.gettempdir
+    ):
+        super().__init__(name, arguments, sceptre_user_data, connection_manager, stack_group_config)
+        self.invoker_class = invoker_class
+        self.get_temp_dir = get_temp_dir
+
+    def schema(self) -> dict:
+        """This is the json schema of the template handler. It is required by Sceptre to define
+        template handler parameters.
+        """
+        return {
+            "type": "object",
+            "properties": {
+                "sam_template_path": {"type": "string"},
+                "artifact_prefix": {"type": "string"},
+                "artifact_bucket_name": {"type": "string"},
+                "sam_build_args": {"type": "object"},
+                "sam_package_args": {"type": "object"}
+            },
+            "required": [
+                "sam_template_path",
+                "artifact_bucket_name",
+            ]
+        }
+
+    def handle(self) -> str:
+        invoker = self.invoker_class(
+            connection_manager=self.connection_manager,
+            sam_directory=self.sam_directory
+        )
+        self._create_generation_destination()
+        self._build(invoker)
+        self._package(invoker)
```

Comment:
hmm, package does not deploy so where is the step to deploy the sam app in the AWS acccount?

Reply:
You don't need SAM to deploy. That's what Sceptre is for! `package` _does_ upload the files to S3, however.

Reply:
To clarify, `sam package` will take the built artifacts, upload them to S3, and transform the uris on the resources to point to the uploaded s3 keys rather than local file paths and outputs that transformed template. The transformed template is what Sceptre will deploy.

Reply:
yeah,  think the thing that might help is to provide an example of a more complete sceptre template that includes the SAM template, stack_name, parameters, etc.. with this description.  I think describing what happens to the transformed (output) template and how it gets passed to sceptre to deploy the SAM app will help clarify what this handler is doing.

Reply:
I've added both a full stack config as the example as well as a whole section on the readme for "how does this handler work"